### ISSUE TITLE
(Discussion) Extended Callbacks

### ISSUE THREAD

**Issue by erfannoury** [fchollet/keras#764]
This issue is a place to discuss more about what callbacks can be created but the current implementation of callbacks and its probable limitations makes creating these callbacks impossible.

One particular case is discussed here https://github.com/fchollet/keras/issues/763.
To be able to reset the hidden state of an RNN not only after some predefined number of epochs, but also inside any mini-batch, requires that the callback have access to the mini-batch data.

What are other use cases? Is this particular case feasible at all?

**Comment by fchollet** [comment]
A common use case would be computing custom metrics to be monitored during training. In general I do believe we need a simple, efficient way to access mini-batch data in callbacks.

**Comment by EderSantana** [comment]
> A common use case would be computing custom metrics to be monitored during training.

I actually needed that before. How about extending the callback methods to optionally get a third input `data`? For the case on `on_batch_*` methods it will be the current batch.

**Comment by fchollet** [comment]
Quoting myself from a while back:

> The (probably) better solution would to make the `ins` arrays and their
text labels accessible to the callbacks, and abstract this away for metric
monitoring by developing a new callback that exposes this system with a
friendly API.

https://github.com/fchollet/keras/issues/626#issuecomment-137503183

Basically the callbacks would be getting per-minibatch inputs, labels, weights.

**Comment by pascalxia** [comment]
Any news on this issue? Is there a solution finally?

**Comment by drewm1980** [comment]
I am also interested in a solution. I would like to log intermediate values of my network in a Callback. Since some backends might (at least I hope) compile out some intermediate tensor values for performance reasons, I understand it will be necessary to-re compile/run the network with eval() on the same input data as feed, but to do that, the actual input data from the last/current minibatch needs to be available to the callback.
